torch/csrc/distributed/rpc/FutureMessage.cpp

```diff
@@ -52,7 +51,7 @@ void FutureMessage::fireCallbacks() {
   // There is no need to protect callbacks with the lock.
   // Once completed_ is set to true, no one can add new callback to the list.
   for (auto& callback : callbacks) {
-    callback(message_);
+    callback();
```

Comment:
When we write something like this: `FutureMessage.then(cb1).then(cb2)`, from a user point of view cb1 should act on the result of FutureMessage and cb2 should act on the result of cb1. In the current version, cb1 and cb2 both act on the result of FutureMessage. Also, FutureMessage.then(cb1) returns the same FutureMessage back. This is confusing for the user, since they might expect that the return value of `FutureMessage.then(cb1)` is a future that will be ready once cb1 is done.

I think the right way to do this is to have FutureMessage.then(cb1) return a new future with one callback. This new future waits for the original future to complete and then runs a single callback. Now, when we add .then(cb2), we create a new future which waits for the future from `FutureMessage.then(cb1)` to complete and runs on callback cb2. I feel this might be more intuitive for users.

Reply:
Make senses! I think it is necessary to have multiple Future object and support `then().then()` for asynchronous UDF. But as that is a non-goal for this PR, let me just change the API to `add_callback()` and disallow chaining, which should be enough to avoid confusion.